File: rb-tree.cpp
#include <iostream>
#include <iomanip>
#include "rb-tree.h"

using namespace std;

void RBtree::insert(int i)
{
    node *n = new node;
    n->key = i;
    n->left = NULL;
    n->right = NULL;

    if(root == NULL)
    {
        root = n;
        n->color = BLACK;
        n->parent = NULL;
    }
    else
    {
        node *p = root;
        node *p_parent = NULL;
        while (p != NULL){
            p_parent = p;
            if(p->key < i){
                p = p->right;
            } else {
                p = p->left;
            }
        }

        n->parent = p_parent;
        n->color = RED;
        if(p_parent->key < n->key) {
            p_parent->right = n;
        } else {
            p_parent->left = n;
        }
    }
    size++;
    insertfix(n);
}

void RBtree::insertfix(node *z)
{
    while(z != root && z->parent->color == RED)
    {
        if(z->parent == z->parent->parent->left)
        {
            node *g = z->parent->parent;
            node *y = g->right;
            if(y !=NULL && y->color == RED)
            {
                z->parent->color = BLACK;
                y->color = BLACK;
                g->color = RED;
                z = g;
            }
            else
            {
                if(z->parent->right == z)
                {
                    z = z->parent;
                    leftrotate(z);
                }

                z->parent->color = BLACK;
                g->color = RED;
                rightrotate(g);
            }
        }
        else
        {
            node *g = z->parent->parent;
            node *y = g->left;

            if(y !=NULL && y->color == RED)
            {
                z->parent->color = BLACK;
                y->color = BLACK;
                g->color = RED;
                z = g;
            }
            else
            {
                if(z->parent->left == z)
                {
                    z = z->parent;
                    rightrotate(z);
                }

                z->parent->color = BLACK;
                g->color = RED;
                leftrotate(g);
            }
        }
    }
    root->color=BLACK;
}

void RBtree::del(int x)
{
     if(root == NULL)
     {
           return ;
     }
     node *p;
     p = root;
     node *y = NULL;
     node *q = NULL;
     int found = 0;
     while(p != NULL && found == 0)
     {
           if(p->key == x)
               found = 1;
               // Where is the size decrease?
           if(found == 0)
           {
                 if(p->key < x)
                    p = p->right;
                 else
                    p = p->left;
           }
     }
     if(found == 0){
        return ;
    }
     else
     {
         if(p->left == NULL || p->right == NULL)
              y = p;
         else
              y = successor(p);
         if(y->left != NULL)
              q = y->left;
         else
         {
              if(y->right != NULL)
                   q = y->right;
              else
                   q = NULL;
         }
         if(q != NULL)
              q->parent = y->parent;
         if(y->parent == NULL)
              root = q;
         else
         {
             if(y == y->parent->left)
                y->parent->left = q;
             else
                y->parent->right = q;
         }
         if(y != p)
         {
             p->color = y->color;
             p->key = y->key;
         }
         if(y->color ==BLACK)
             delfix(q);
        size--;
     }
}

void RBtree::delfix(node *p)
{
    node *s;
    while(p != NULL && p != root && p->color == BLACK)
    {
        if(p->parent->left == p){
            s = p->parent->right;
            if(s != NULL && s->color == RED){
                s->color = BLACK;
                p->parent->color = RED;
                leftrotate(p->parent);
                s = p->parent->right;
            }
            if(s == NULL){
                break;
            }
            if((s->right == NULL || s->right->color == BLACK) && (s->left == NULL || s->left->color == BLACK)){
                s->color = RED;
                p = p->parent;
            } else {
                if(s->right == NULL || s->right->color == BLACK) {
                    if (s->left != NULL )
                        s->left->color = BLACK;
                    s->color = RED;
                    rightrotate(s);
                    s = p->parent->right;
                }
                s->color = p->parent->color;
                p->parent->color = BLACK;
                s->right->color = BLACK;
                leftrotate(p->parent);
                p=root;
            }
        } else {
            s = p->parent->left;
            if(s != NULL && s->color == RED) {
                s->color = BLACK;
                p->parent->color = RED;
                rightrotate(p->parent);
                s = p->parent->left;
            }
            if(s == NULL){
                break;
            }
            if((s->left == NULL || s->left->color == BLACK) && (s->right == NULL || s->right->color == BLACK))
            {
                s->color = RED;
                p = p->parent;
            } else {
                if(s->left == NULL || s->left->color == BLACK) {
                    if (s->right != NULL)
                        s->right->color = BLACK;
                    s->color = RED;
                    leftrotate(s);
                    s = p->parent->left;
                }
                s->color = p->parent->color;
                p->parent->color = BLACK;
                if(s->left != NULL)
                    s->left->color = BLACK;
                rightrotate(p->parent);
                p = root;
            }
        }
        p->color = BLACK;
        root->color = BLACK;
    }
}

void RBtree::leftrotate(node *p)
{
    if(p->right == NULL)
        return ;
    else
    {
        node *y = p->right;
        if(y->left != NULL){
            p->right = y->left;
            y->left->parent = p;
        } else {
            p->right = NULL;
        }
        if(p->parent != NULL){
            y->parent = p->parent;
        }
        if(p->parent == NULL){
            root = y;
            y->parent = NULL;
        } else {
            if(p == p->parent->left){
                p->parent->left = y;
            } else{
                p->parent->right = y;
            }
        }
        y->left = p;
        p->parent = y;
    }
}
void RBtree::rightrotate(node *p)
{
    if(p->left == NULL)
        return ;
    else
    {
        node *y = p->left;
        if(y->right != NULL) {
            p->left = y->right;
            y->right->parent = p;
        } else {
            p->left = NULL;
        }
        if(p->parent != NULL) {
            y->parent = p->parent;
        }
        if(p->parent == NULL){
            root = y;
            y->parent = NULL;
        } else {
            if(p == p->parent->left){
                p->parent->left = y;
            } else {
                p->parent->right = y;
            }
        }
        y->right = p;
        p->parent = y;
    }
}

node* RBtree::successor(node *p)
{
    node *y = NULL;
    if(p->left != NULL){
        y = p->right;
        while(y->left != NULL)
            y = y->left;
    }
    else{
        y = p->left;
        while(y->right != NULL)
            y = y->right;
    }
    return y;
}

bool RBtree::search(int x)
{
    if(root == NULL)  return false;
    node *p = root;
    bool found = false;
    while(p != NULL && !found)
    {
        if(p->key == x)
            return true;
        if(!found)
        {
            if(p->key > x){
                p = p->left;
            } else {
                p = p->right;
            }
        }
    }
    return false;
}

void postorder(node* p, int indent)
{
    if(p != NULL) {
        if(p->left) postorder(p->left, indent+4);
        if(p->right) postorder(p->right, indent+4);
        if (indent) {
            std::cout << std::setw(indent) << ' ';
        }
        cout<< p->key << "\n ";
    }
}

File: rb-tree.lua
Node = {}
Node.__index = Node

function Node.new(key)
    local node = {}
    setmetatable(node, Node)
    node.key = key
    node.parent = nil
    node.color = nil
    node.left = nil
    node.right = nil
    return node
end

RBtree = {}
RBtree.__index = RBtree

function RBtree.new()
    local tree = {}
    setmetatable(tree, RBtree)
    tree.root = nil
    tree.size = 0
    return tree
end

function RBtree:insert(i)
    local n = Node:new()
    n.key = i
    n.left = nil
    n.right = nil
    if self.root == nil then
        self.root = n
        n.color = 'b'
        n.parent = nil
    else
        local p = self.root
        local parent = nil
        while p ~= nil do
            parent = p
            if p.key < i then
                p = p.right
            else
                p = p.left
            end
        end
        
        p = parent
        n.parent = p
        n.color = 'r'
        if p.key < n.key then
            p.right = n
        else
            p.left = n
        end
    end
    self.size = self.size + 1
    self:insertfix(n)
end

function RBtree:insertfix(z)
    while z ~= self.root and z.parent.color == 'r' do
        if z.parent == z.parent.parent.left then
            local g = z.parent.parent
            local y = g.right
            if y ~= nil and y.color == 'r' then
                z.parent.color = 'b'
                y.color = 'b'
                g.color = 'r'
                z = g
            else
                if z.parent.right == z then
                    z = z.parent
                    self:leftrotate(z)
                end

                z.parent.color = 'b'
                g.color = 'r'
                self:rightrotate(g)
            end
        else
            local g = z.parent.parent
            local y = g.left
            if y ~= nil and y.color == 'r' then
                z.parent.color = 'b'
                y.color = 'b'
                g.color = 'r'
                z = g
            else
                if z.parent.left == z then
                    z = z.parent
                    self:rightrotate(z)
                end

                z.parent.color = 'b'
                g.color = 'r'
                self:leftrotate(g)
            end
        end
    end
    self.root.color = 'b'
end

function RBtree:del(i)
    if self.root == nil then return end
    
    local p = self.root
    local y = nil
    local q = nil
    local found = false

    while p ~= nil and not found do
        if p.key == i then
            found = true
            self.size = self.size - 1
        end

        if not found then
            if p.key < i then
                p = p.right
            else
                p = p.left
            end
        end
    end

    if not found then return end

    if p.left == nil or p.right == nil then
        y = p
    else
        y = self:successor(p)
    end
    if y.left ~= nil then
        q = y.left
    else
        if y.right ~= nil then
            q = y.right
        else
            q = nil
        end
    end
    if q ~= nil then
        q.parent = y.parent
    end
    if y.parent == nil then
        self.root = q
    else
        if y == y.parent.left then
            y.parent.left = q
        else
            y.parent.right = q
        end
    end
    if y ~= p then
        p.color = y.color
        p.key = y.key
    end
    if y.color == 'b' then
        self:delfix(q)
    end
end

function RBtree:delfix(p)
    if p == nil then
        return
    end
    
    local s = nil
    while p~=nil and p ~= self.root and p.color == 'b' do
        if p.parent.left == p then
            s = p.parent.right
            if s~=nil and s.color == 'r' then
                s.color = 'b'
                p.parent.color = 'r'
                self:leftrotate(p.parent)
                s = p.parent.right
            end
            if s==nil then
                break
            end
            if (s.right == nil or s.right.color == 'b') and (s.left == nil or s.left.color == 'b') then
                s.color = 'r'
                p = p.parent
            else
                if s.right == nil or s.right.color == 'b' then
                    if s.left ~= nil then
                        s.left.color = 'b'
                    end
                    s.color = 'r'
                    self:rightrotate(s)
                    s = p.parent.right
                end
                s.color = p.parent.color
                p.parent.color = 'b'
                s.right.color = 'b'
                self:leftrotate(p.parent)
                p = self.root
            end
        else
            s = p.parent.left
            if s ~= nil and s.color == 'r' then
                s.color = 'b'
                p.parent.color = 'r'
                self:rightrotate(p.parent)
                s = p.parent.left
            end
            if s==nil then
                break
            end
            if (s.left == nil or s.left.color == 'b') and (s.right == nil or s.right.color == 'b') then
                s.color = 'r'
                p = p.parent
            else
                if s.left == nil or s.left.color == 'b' then
                    if s.right~= nil then
                        s.right.color = 'b'
                    end
                    s.color = 'r'
                    self:leftrotate(s)
                    s = p.parent.left
                end
                s.color = p.parent.color
                p.parent.color = 'b'
                if s.left ~= nil then
                    s.left.color = 'b'
                end
                self:rightrotate(p.parent)
                p = self.root
            end
        end
        p.color = 'b'
        self.root.color = 'b'
    end
end

function RBtree:leftrotate(p)
    if p.right == nil then
        return
    else
        local y = p.right
        if y.left ~= nil then
            p.right = y.left
            y.left.parent = p
        else
            p.right = nil
        end
        if p.parent ~= nil then
            y.parent = p.parent
        end
        if p.parent == nil then
            self.root = y
            self.root.parent = nil
        else
            if p == p.parent.left then
                p.parent.left = y
            else
                p.parent.right = y
            end
        end
        y.left = p
        p.parent = y
    end

end

function RBtree:rightrotate(p)
    if p.left == nil then
        return
    else
        local y = p.left
        if y.right ~= nil then
            p.left = y.right
            y.right.parent = p
        else
            p.left = nil
        end
        if p.parent ~= nil then
            y.parent = p.parent
        end
        if p.parent == nil then
            self.root = y
            y.parent = nil -- fix
        else
            if p == p.parent.left then
                p.parent.left = y
            else
                p.parent.right = y
            end
        end
        y.right = p
        p.parent = y
    end
end

function RBtree:successor(p)
    local y = nil
    if p.left ~= nil then
        y = p.right
        while y.left ~= nil do
            y = y.left
        end
    else
        y = p.left
        while y.right ~= nil do
            y = y.right
        end
    end
    return y
end

function RBtree:search(x)
    if self.root == nil then
        return false
    end
    local p = self.root
    local found = false
    while p ~= nil and not found do
        if p.key == x then
            return true
        end
        if not found then
            if p.key < x then
                p = p.right
            else
                p = p.left
            end
        end
    end
    return false
end